src/cargo/ops/cargo_add/mod.rs

```diff
@@ -518,28 +532,44 @@ fn populate_available_features(
     }
 
     let query = dependency.query(config)?;
-    let possibilities = loop {
-        match registry.query_vec(&query, true) {
-            std::task::Poll::Ready(res) => {
-                break res?;
+    match query {
+        MaybeWorkspace::Workspace(workspace) => {
+            if let Some(features) = workspace.features {
+                let mut b_tree: BTreeMap<String, Vec<String>> = BTreeMap::new();
+                for feature in features {
+                    b_tree.insert(feature, vec![]);
+                }
+                dependency = dependency.set_available_features(b_tree)
```

Comment:
Don't we need to be looking up the workspace's dependency and getting its query?

Reply:
Possibly, this was one area that I wasn't sure how we should be going about this. I know that this makes it so `foo = { workspace = true, features = ["test"]}` works.

Reply:
We have two feature lists
- What the user specified
- What the crate defines

This code is putting the user-specified features into the list of what the crate defines so anything the user enters will always be valid.